### PR TITLE
Changed/Fixed enemy jeep behaviour

### PR DESCRIPTION
**Pull request by TokyoSU**
## todo

- [X] I have added a changelog entry to CHANGELOG.md file on the branch/fork (if it is an internal change then it is not needed) 
- [X] Pull request meets the Coding Conventions standards: https://github.com/MontyTRC89/TombEngine/blob/master/CONTRIBUTING.md#coding-conventions

## Links to issue(s) this pull request concerns (if applicable)

"n/a"

## Description of pull request 

Check changelogs.md -> Enemy jeep.
Also corrected the changelogs a bit.

File required/for testing:
[KV5_jeeptest.zip](https://github.com/user-attachments/files/16603144/KV5_jeeptest.zip)

Enemy Jeep (1.5)

The enemy jeep from TR4 has been re-implemented into TombEngine but the setup has changed. It requires the use of AI_X1 for path (need to start at 0 and end at whatever your path need to end) and AI_X2 objects to change it's behaviour through this path, AI_X2 ocb:

OCB 1: Drop grenade.
OCB 2: Vault.
OCB 3: Jump over pit.
OCB 4: Wait until player is near.
OCB 5: Disappear.
OCB 6: Activate heavy trigger.
OCB 7: Another vault (Need animation + state setup).
OCB 8: Another jump pit (Need animation + state setup).

Furthermore: you set the OCB on the ENEMY_JEEP object itself for the desired effect:

OCB 0: Will wait until the player is on a vehicle.
OCB 1: Will start regardless of whether the player is on a vehicle or not.

## Warning:
Enemy jeep have changed drasticaly, the original object is not valid anymore !

### CODE REVIEW

File: TombEngine/Game/Setup.h [hunk 1]
```diff
@@ -39,10 +39,11 @@ enum class LotType
...
 	HumanPlusJumpAndMonkey,
 	Flyer,
-	Blockable,	  // For large creatures such as trex and shiva.
-	Spider,		  // Only 2 block vault allowed.
-	Ape,		  // Only 2 block vault allowed.
-	SnowmobileGun // Only 1 block vault allowed and 4 block drop max.
+	Blockable,	   // For large creatures such as trex and shiva.
+	Spider,		   // Only 2 block vault allowed.
+	Ape,		   // Only 2 block vault allowed.
+	SnowmobileGun, // Only 1 block vault allowed and 4 block drop max.
```
Comment (Stranger1992) [inline TombEngine/Game/Setup.h:45]: Indent comment to be inline with the rest.

File: TombEngine/Game/Setup.h [hunk 2]
```diff
@@ -39,10 +39,11 @@ enum class LotType
...
+	EnemyJeep,
```
Comment (Stranger1992) [inline TombEngine/Game/Setup.h:46]: Could a comment be written explicitly explaining why the ENEMY_JEEP is different?

File: TombEngine/Objects/TR4/tr4_objects.cpp [hunk 3]
```diff
@@ -12,7 +12,7 @@
 
 // Creatures
 #include "Objects/TR4/Entity/Wraith.h" // OFF
-#include "Objects/TR4/Entity/tr4_enemy_jeep.h"
+#include "Objects/TR4/Entity/EnemyJeep.h" // WIP
```
Comment (Stranger1992) [inline TombEngine/Objects/TR4/tr4_objects.cpp:15]: Remove comment

File: TombEngine/Objects/TR4/Entity/EnemyJeep.cpp [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+		// Make the enemy jeep start directly instead of waiting the player to enter a vehicle.
+		EJ_NO_PLAYER_VEHICLE_REQUIRED = 1
+	};
+
+	enum EnemyJeepX2Ocb
+	{
+		X2_DROP_GRENADE = 1,
+		X2_DROP = 2,
+		X2_JUMP_PIT = 3,
+		// Need ocb 4 + block distance
```
Comment (Stranger1992) [inline TombEngine/Objects/TR4/Entity/EnemyJeep.cpp:84]: This comment isnt clear if it has been implemented or still to work on ?

File: TombEngine/Objects/TR4/Entity/EnemyJeep.cpp [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+	constexpr auto ENEMY_JEEP_PITCH_MAX = 120.0f;
+	constexpr auto ENEMY_JEEP_PITCH_WHEEL_SPEED_MULTIPLIER = 12.0f;
+	constexpr auto ENEMY_JEEP_WHEEL_LEFTRIGHT_TURN_MINIMUM = ANGLE(12.0f);
+
+	const auto EnemyJeepGrenadeBite = CreatureBiteInfo(Vector3(0.0f, -640.0f, -768.0f), ENEMY_JEEP_CENTER_MESH);
+	const auto EnemyJeepRightLightBite = CreatureBiteInfo(Vector3(200.0f, -144.0f, -768.0f), ENEMY_JEEP_CENTER_MESH);
+	const auto EnemyJeepLeftLightBite = CreatureBiteInfo(Vector3(-200.0f, -144.0f, -768.0f), ENEMY_JEEP_CENTER_MESH);
+
+	/// <summary>
+	/// Enable/Disable the light mesh that are on the back of the jeep.
```
Comment (Stranger1992) [inline TombEngine/Objects/TR4/Entity/EnemyJeep.cpp:111]: "Enable/Disable the light mesh on the jeep's back".

**Review by Stranger1992 — CHANGES_REQUESTED** [review]
Tested in-game:

## ENEMY_JEEP:

* OCB 0: Working
* OCB 1: Working

## AI_X1

* OCB working. Path is being followed nicely. However: if there are plans for multiple enemy jeeps: then this may need to be revised? I Have not tested what happens if you change the OCB in run-time (for example if the player does something in game that will change the jeep's path).

## AI_X2

* OCB 1: Working. Fires grenades as expected
* OCB 2: Working. Clarity necessary as to what values to be changed for different fall distances though. 
* OCB 3: Working. Clarity necessary as to what values to be changed for different jump distances though. 
* OCB 4: Working
* OCB 5: Working
* OCB 6: Working
* OCB 7: Unsure what this does so not tested
* OCB 8: Same as OCB 7.

File: TombEngine/Game/Setup.cpp [hunk 6]
```diff
@@ -79,7 +79,14 @@ ObjectInfo& ObjectHandler::GetFirstAvailableObject()
 // NOTE: JointRotationFlags allows bones to be rotated with CreatureJoint().
 void ObjectInfo::SetBoneRotationFlags(int boneID, int flags)
 {
-	g_Level.Bones[boneIndex + (boneID * 4)] |= flags;
+	int boneIndex = boneIndex + (boneID * 4);
```
Comment (Lwmte) [inline TombEngine/Game/Setup.cpp:82]: `boneIndex` does not exist. This line results in a compiler error.

File: TombEngine/Game/control/box.cpp [hunk 7]
```diff
@@ -1432,69 +1437,119 @@ void FindAITarget(CreatureInfo* creature, short objectNumber)
 	}
 }
 
-void FindAITargetObject(CreatureInfo* creature, int objectNumber)
+void FindAITargetObject(ItemInfo& item, GAME_OBJECT_ID objectID)
```
Comment (Lwmte) [inline TombEngine/Game/control/box.cpp:1440]: Why function argument was changed from `CreatureInfo` to `ItemInfo`?

File: TombEngine/Game/control/box.h [hunk 8]
```diff
@@ -83,10 +83,24 @@ constexpr auto CLIP_ALL    = (CLIP_LEFT | CLIP_RIGHT | CLIP_TOP | CLIP_BOTTOM);
 
 constexpr auto CLIP_SECONDARY = 0x10;
 
+struct AITargetFlags
```
Comment (Lwmte) [inline TombEngine/Game/control/box.h:86]: Why it is called `AITargetFlags`? It should be called `AITargetInfo`.

File: TombEngine/Game/effects/effects.cpp [hunk 9]
```diff
@@ -1255,7 +1255,7 @@ void TriggerDynamicLight(const Vector3& pos, const Color& color, float falloff)
 		color.x * UCHAR_MAX, color.y * UCHAR_MAX, color.z * UCHAR_MAX);
 }
 
-// Deprecated. Use above version instead.
+// DEPRECATED! DO NOT USE!!
```
Comment (Lwmte) [inline TombEngine/Game/effects/effects.cpp:1258]: ?

Comment (Jakub768) [reply TombEngine/Game/effects/effects.cpp:1258]: If it is deprecated and do not use, why not just delete it?

File: TombEngine/Game/control/box.cpp [hunk 10]
```diff
@@ -1432,69 +1438,101 @@ void FindAITarget(CreatureInfo* creature, short objectNumber)
...
+		return false;
 
-	aiItem.ObjectNumber = foundObject->objectNumber;
-	aiItem.RoomNumber = foundObject->roomNumber;
-	aiItem.Pose.Position = foundObject->pos.Position;
-	aiItem.Pose.Orientation.y = foundObject->pos.Orientation.y;
-	aiItem.Flags = foundObject->flags;
-	aiItem.TriggerFlags = foundObject->triggerFlags;
-	aiItem.BoxNumber = foundObject->boxNumber;
+	auto aiItem = ItemInfo{};
```
Comment (Lwmte) [inline TombEngine/Game/control/box.cpp:1519]: Why new `ItemInfo` is created and returned by this function?

File: TombEngine/Game/control/box.h (hunk 8 above)
Comment (TokyoSU) [reply TombEngine/Game/control/box.h]: Will change.

File: TombEngine/Game/control/box.cpp (hunk 7 above)
Comment (TokyoSU) [reply TombEngine/Game/control/box.cpp]: Sezz did it :x

File: TombEngine/Game/Setup.h [hunk 11]
```diff
@@ -39,10 +39,11 @@ enum class LotType
...
 	Flyer,
-	Blockable,	  // For large creatures such as trex and shiva.
-	Spider,		  // Only 2 block vault allowed.
-	Ape,		  // Only 2 block vault allowed.
-	SnowmobileGun // Only 1 block vault allowed and 4 block drop max.
+	Blockable,	   // For large creatures such as trex and shiva.
+	Spider,		   // Only 2 block vault allowed.
+	Ape,		   // Only 2 block vault allowed.
+	SnowmobileGun, // Only 1 block vault allowed and 4 block drop max.
+	EnemyJeep
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply TombEngine/Game/Setup.h:46]: No because i need more click that the other offer me.

File: TombEngine/Game/effects/effects.cpp (hunk 9 above)
Comment (TokyoSU) [reply TombEngine/Game/effects/effects.cpp]: Sezz did that, don't know why this function is used everywhere :x

File: TombEngine/Game/control/lot.cpp [hunk 12]
```diff
@@ -197,6 +197,13 @@ void InitializeSlot(short itemNumber, bool makeTarget)
 			creature->LOT.Drop = -BLOCK(1);
 			creature->LOT.Zone = ZoneType::Human;
 			break;
+
+		case LotType::EnemyJeep:
+			creature->LOT.Step = BLOCK(4);
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply TombEngine/Game/control/lot.cpp:202]: To avoid being blocked when going up/down

File: TombEngine/Game/control/box.cpp (hunk 10 above)
Comment (TokyoSU) [reply TombEngine/Game/control/box.cpp:1519]: It's required if i want the enemy to move toward it, creature info expect a ItemInfo* in FindAITargetObject()

File: TombEngine/Game/control/lot.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (TokyoSU) [reply TombEngine/Game/control/lot.cpp:202]: While change to BLOCK(1.5f)

**Pull request merged**